_applications/equoid.adoc

```diff
@@ -0,0 +1,206 @@
+    -p SPARK_OPTIONS='--driver-java-options=-Dvertx.cacheDirBase=/tmp'
+....
+
+=== Launch Web UI and Set Up Keycloak
+
+Equoid's web UI is launched using a script developed by one of Equoid's contributors, Jiri Kremser. The following calls will utilize that script to set up the UI as well as a Keycloak instance for users and login. Finally, an edit role is given to the project's service account so additional data handler instances can be created/modified. 
+
+....
+BASE_URL="https://raw.githubusercontent.com/radanalyticsio/equoid-ui/master/ocp/"
+curl -sSL $BASE_URL/ocp-apply.sh | \
+    BASE_URL="$BASE_URL" \
+    KC_REALM_PATH="web-ui/keycloak/realm-config" \
+    bash -s stable
+oc policy add-role-to-user edit system:serviceaccount:`oc project -q`:default
+....
+
+[[usage]]
+== Usage
+
+=== Web Interface
+With all the microservices running, you can now view a visualization of the data stream's top-ks for the parameters you set. By browsing to the OpenShift webui for your cluster, logging in, and looking at your project's overview, you should see a pod/application named equoid-ui with an associated URI. By clicking on the latter, you will then be at the login screen for your equoid system. Log in using either of the default credential sets listed on the login page, then click on the graph tab. You should see a visualization of your stream's data which updates periodically. 
+
+In addition, you can add a frequent item by clicking on the add item button. Finally, you can add another data handler via the command line, a script is included in equoid-openshift to facilitate additional data handler instances. Once these are up and running, the web UI will automatically add their data in a separate graph tab. 
+ 
+=== Equoid Application Settings
+We provide a number of variables to modify per the needs of end users. In this section, these are described. 
+
+==== Data Publisher
+
+`DATA_URL_PRIMARY` - URL of either full dataset or list of all field values for the linear and single `OP_MODE` settings, respectively.
+
+`DATA_URL_SECONDARY` - URL of all field values for the dual `OP_MODE` setting.
+
+`OP_MODE` - Operating mode: single for a list of field values from which random elements will be generated, dual for two lists of field values from which random elements will be generated, linear for a dataset meant to be read in sequence and transmitted.
+
+==== Data Handler
+
+`WINDOW_SECONDS` - Size, in seconds of window for which the top-k elements should be determined. 
+
+`SLIDE_SECONDS` - Size, in seconds, of the amount to slide the sample window by each iteration. 
+
+`BATCH_SECONDS` - Size, in seconds, of the batch size to be acquired from the broker.
+
+`OP_MODE` - As with the data publisher, single for a list of field values from which random elements will be generated, dual for two lists of field values from which random elements will be generated, linear for a dataset meant to be read in sequence and transmitted.
```

Comment:
why handler should know about the method how the data were ingested? I checked the code so I guess, I got the point of the modes, but 


I guess, there is a bug on this line 

https://github.com/radanalyticsio/equoid-data-handler/blob/master/src/main/scala/io/radanalytics/equoid/DataHandler.scala#L33

"dual" behaves the same as "linear" + single ignores second item (even though there should be none according to publisher logic, the comma should be there only in case of "dual" mode: https://github.com/radanalyticsio/equoid-data-publisher/blob/master/src/main/scala/io/radanalytics/equoid/DataPublisher.scala#L108)

I guess, I am totally missing the point of the operation mode. Why isn't it enough to create two publishers each with its own dataset? Is this all about the preparation for the "joint" count-min-sketches? If yes, wouldn't it be better to have some better format for the events than just text. But for instance some array like `["foo1", "foo2", "42", ..]` with variable length and data handler could do the job for each element in the array? Well, transmitting the elements 1-by-1 would do the same, I still can't see the purpose.

I've noticed, you are passing the opMode to the amqp code (`AMQPUtils.createStream(..)`), what does AMQP does with this mode?

Sorry, I don't want to complicate this PR, we can chat about it.

Reply:
OP_MODE is meant to illustrate how one can add functionality to the system, but admittedly is not as fleshed out as one might do in production. The idea I envisaged is that one can specify a new OP_MODE to suit their needs, which may/may not have a impact on both the publisher and handler.

Perhaps I could add a blurb about this being an example but not necessarily how one would do it in production?